{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Convert VCFs to Pandas DataFrames and save to .pkl format. Convert snpAD-called heterozygotes to homozygous calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chimp\n",
      "HG02666\n",
      "HG02982\n",
      "HGDP00056\n",
      "HGDP00057\n",
      "HGDP00103\n",
      "HGDP00127\n",
      "HGDP00136\n",
      "HGDP00213\n",
      "HGDP00218\n",
      "HGDP00453\n",
      "HGDP00475\n",
      "HGDP01200\n",
      "HGDP01214\n",
      "HGDP01225\n",
      "HGDP01298\n",
      "HGDP01406\n",
      "HGDP00478\n",
      "HGDP00549\n",
      "HGDP00757\n",
      "HGDP00908\n",
      "HGDP00931\n",
      "HGDP00984\n",
      "HGDP01009\n",
      "HGDP01029\n",
      "HGDP01031\n",
      "HGDP01190\n",
      "HGDP01192\n",
      "A00\n",
      "A00_I10871\n",
      "Mezmaiskaya2\n",
      "Chagyrskaya2\n",
      "Ust_Ishim\n",
      "Yana1\n",
      "Loschbour\n",
      "HG02666\n",
      "HG02982\n",
      "HGDP00056\n",
      "HGDP00057\n",
      "HGDP00103\n",
      "HGDP00127\n",
      "HGDP00136\n",
      "HGDP00213\n",
      "HGDP00218\n",
      "HGDP00453\n",
      "HGDP00475\n",
      "HGDP01200\n",
      "HGDP01214\n",
      "HGDP01225\n",
      "HGDP01298\n",
      "HGDP01406\n",
      "HGDP00478\n",
      "HGDP00549\n",
      "HGDP00757\n",
      "HGDP00908\n",
      "HGDP00931\n",
      "HGDP00984\n",
      "HGDP01009\n",
      "HGDP01029\n",
      "HGDP01031\n",
      "HGDP01190\n",
      "HGDP01192\n",
      "A00\n",
      "A00_I10871\n",
      "Mezmaiskaya2\n",
      "Chagyrskaya2\n",
      "Ust_Ishim\n",
      "Yana1\n",
      "Loschbour\n",
      "chimp\n",
      "A00_I10871\n",
      "HG02666\n",
      "HG02982\n",
      "HGDP00056\n",
      "HGDP00057\n",
      "HGDP00103\n",
      "HGDP00127\n",
      "HGDP00136\n",
      "HGDP00213\n",
      "HGDP00218\n",
      "HGDP00453\n",
      "HGDP00475\n",
      "HGDP00478\n",
      "HGDP00549\n",
      "HGDP00757\n",
      "HGDP00908\n",
      "HGDP00931\n",
      "HGDP00984\n",
      "HGDP01009\n",
      "HGDP01029\n",
      "HGDP01031\n",
      "HGDP01190\n",
      "HGDP01192\n",
      "HGDP01200\n",
      "HGDP01214\n",
      "HGDP01225\n",
      "HGDP01298\n",
      "HGDP01406\n",
      "A00_I10871\n",
      "HG02666\n",
      "HG02982\n",
      "HGDP00056\n",
      "HGDP00057\n",
      "HGDP00103\n",
      "HGDP00127\n",
      "HGDP00136\n",
      "HGDP00213\n",
      "HGDP00218\n",
      "HGDP00453\n",
      "HGDP00475\n",
      "HGDP00478\n",
      "HGDP00549\n",
      "HGDP00757\n",
      "HGDP00908\n",
      "HGDP00931\n",
      "HGDP00984\n",
      "HGDP01009\n",
      "HGDP01029\n",
      "HGDP01031\n",
      "HGDP01190\n",
      "HGDP01192\n",
      "HGDP01200\n",
      "HGDP01214\n",
      "HGDP01225\n",
      "HGDP01298\n",
      "HGDP01406\n",
      "chimp\n",
      "A00_I10871\n",
      "HG02666\n",
      "HG02982\n",
      "HGDP00056\n",
      "HGDP00057\n",
      "HGDP00103\n",
      "HGDP00127\n",
      "HGDP00136\n",
      "HGDP00213\n",
      "HGDP00218\n",
      "HGDP00453\n",
      "HGDP00475\n",
      "HGDP00478\n",
      "HGDP00549\n",
      "HGDP00757\n",
      "HGDP00908\n",
      "HGDP00931\n",
      "HGDP00984\n",
      "HGDP01009\n",
      "HGDP01029\n",
      "HGDP01031\n",
      "HGDP01190\n",
      "HGDP01192\n",
      "HGDP01200\n",
      "HGDP01214\n",
      "HGDP01225\n",
      "HGDP01298\n",
      "HGDP01406\n",
      "A00_I10871\n",
      "HG02666\n",
      "HG02982\n",
      "HGDP00056\n",
      "HGDP00057\n",
      "HGDP00103\n",
      "HGDP00127\n",
      "HGDP00136\n",
      "HGDP00213\n",
      "HGDP00218\n",
      "HGDP00453\n",
      "HGDP00475\n",
      "HGDP00478\n",
      "HGDP00549\n",
      "HGDP00757\n",
      "HGDP00908\n",
      "HGDP00931\n",
      "HGDP00984\n",
      "HGDP01009\n",
      "HGDP01029\n",
      "HGDP01031\n",
      "HGDP01190\n",
      "HGDP01192\n",
      "HGDP01200\n",
      "HGDP01214\n",
      "HGDP01225\n",
      "HGDP01298\n",
      "HGDP01406\n"
     ]
    }
   ],
   "source": [
    "import utils.VcfHandler as VcfHandler\n",
    "\n",
    "hg19_mapped = VcfHandler.convert_vcf_to_dataframe('data/input_data/hg19_mapped/data_full_merged.hg19_mapped.map35_100.GCcov.noSimpleRepeat.capture_full.vcf.gz', convert_hets=True)\n",
    "hg19_mapped.to_pickle('data/input_data/data_full_merged.hg19_mapped.vcf.gz.pkl')\n",
    "\n",
    "a0b_mapped = VcfHandler.convert_vcf_to_dataframe('data/input_data/A0b_mapped/data_full_merged.A0b_mapped.map35_100.GCcov.noSimpleRepeat.vcf.gz', convert_hets=True)\n",
    "a0b_mapped.to_pickle('data/input_data/data_full_merged.A0b_mapped.vcf.gz.pkl')\n",
    "\n",
    "t2t_mapped = VcfHandler.convert_vcf_to_dataframe('data/input_data/T2T_mapped/data_full_merged.T2T_mapped.map35_100.GCcov.noSimpleRepeat.vcf.gz', convert_hets=True)\n",
    "t2t_mapped.to_pickle('data/input_data/data_full_merged.T2T_mapped.vcf.gz.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate coverage table (based on hg19_mapped VCF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Individual</th>\n",
       "      <th>Coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chagyrskaya2</td>\n",
       "      <td>46.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mezmaiskaya2</td>\n",
       "      <td>15.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A00_I10871</td>\n",
       "      <td>8.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A00</td>\n",
       "      <td>17.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HG02982</td>\n",
       "      <td>14.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HG02666</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HGDP01029</td>\n",
       "      <td>15.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HGDP01406</td>\n",
       "      <td>14.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HGDP00931</td>\n",
       "      <td>15.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HGDP00478</td>\n",
       "      <td>16.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HGDP00453</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HGDP00475</td>\n",
       "      <td>15.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HGDP00984</td>\n",
       "      <td>16.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HGDP00757</td>\n",
       "      <td>15.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HGDP01214</td>\n",
       "      <td>14.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HGDP01031</td>\n",
       "      <td>14.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HGDP01200</td>\n",
       "      <td>14.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HGDP00908</td>\n",
       "      <td>14.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HGDP00103</td>\n",
       "      <td>14.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>HGDP00213</td>\n",
       "      <td>15.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Loschbour</td>\n",
       "      <td>10.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HGDP00127</td>\n",
       "      <td>15.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HGDP00057</td>\n",
       "      <td>13.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>HGDP00056</td>\n",
       "      <td>13.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Ust_Ishim</td>\n",
       "      <td>20.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>HGDP00549</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Yana1</td>\n",
       "      <td>13.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>HGDP01009</td>\n",
       "      <td>16.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>HGDP00136</td>\n",
       "      <td>13.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HGDP00218</td>\n",
       "      <td>15.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>HGDP01298</td>\n",
       "      <td>14.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>HGDP01192</td>\n",
       "      <td>14.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>HGDP01225</td>\n",
       "      <td>13.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>HGDP01190</td>\n",
       "      <td>15.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Individual  Coverage\n",
       "0   Chagyrskaya2      46.8\n",
       "1   Mezmaiskaya2      15.8\n",
       "2     A00_I10871       8.5\n",
       "3            A00      17.2\n",
       "4        HG02982      14.3\n",
       "5        HG02666      15.0\n",
       "6      HGDP01029      15.2\n",
       "7      HGDP01406      14.4\n",
       "8      HGDP00931      15.8\n",
       "9      HGDP00478      16.5\n",
       "10     HGDP00453      15.0\n",
       "11     HGDP00475      15.2\n",
       "12     HGDP00984      16.5\n",
       "13     HGDP00757      15.9\n",
       "14     HGDP01214      14.9\n",
       "15     HGDP01031      14.7\n",
       "16     HGDP01200      14.9\n",
       "17     HGDP00908      14.7\n",
       "18     HGDP00103      14.8\n",
       "19     HGDP00213      15.2\n",
       "20     Loschbour      10.5\n",
       "21     HGDP00127      15.2\n",
       "22     HGDP00057      13.5\n",
       "23     HGDP00056      13.7\n",
       "24     Ust_Ishim      20.0\n",
       "25     HGDP00549      14.0\n",
       "26         Yana1      13.8\n",
       "27     HGDP01009      16.4\n",
       "28     HGDP00136      13.6\n",
       "29     HGDP00218      15.4\n",
       "30     HGDP01298      14.5\n",
       "31     HGDP01192      14.1\n",
       "32     HGDP01225      13.9\n",
       "33     HGDP01190      15.5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import utils.VcfHandler as VcfHandler\n",
    "import utils.YChrDataset as YChrDataset\n",
    "\n",
    "hg19_mapped = pd.read_pickle('data/input_data/data_full_merged.hg19_mapped.vcf.gz.pkl')\n",
    "\n",
    "cov = []\n",
    "for ind in YChrDataset.samples:\n",
    "    coverage = hg19_mapped[ind, 'DP'].mean()\n",
    "    cov.append([ind, round(coverage, 1)])\n",
    "\n",
    "df = pd.DataFrame(cov, columns=['Individual', 'Coverage'])\n",
    "\n",
    "with pd.option_context('display.precision', 1):\n",
    "    display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create fasta files from VCFs using only variable sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import utils.VcfHandler as VcfHandler\n",
    "import utils.YChrDataset as YChrDataset\n",
    "import os\n",
    "\n",
    "if not os.path.exists('data/output_data'):\n",
    "    os.makedirs('data/output_data')\n",
    "\n",
    "hg19_mapped = pd.read_pickle('data/input_data/data_full_merged.hg19_mapped.vcf.gz.pkl')\n",
    "hg19_mapped = VcfHandler.get_genotypes_from_vcf_dataframe(hg19_mapped)\n",
    "\n",
    "VcfHandler.convert_vcf_dataframe_to_fasta_file(hg19_mapped, 'data/output_data/data_full_merged.hg19_mapped.fa', variant_sites_only=True, order=['chimp'] + YChrDataset.samples)\n",
    "\n",
    "a0b_mapped = pd.read_pickle('data/input_data/data_full_merged.A0b_mapped.vcf.gz.pkl')\n",
    "a0b_mapped = VcfHandler.get_genotypes_from_vcf_dataframe(a0b_mapped)\n",
    "\n",
    "VcfHandler.convert_vcf_dataframe_to_fasta_file(a0b_mapped, 'data/output_data/data_full_merged.A0b_mapped.fa', variant_sites_only=True)\n",
    "\n",
    "t2t_mapped = pd.read_pickle('data/input_data/data_full_merged.T2T_mapped.vcf.gz.pkl')\n",
    "t2t_mapped = VcfHandler.get_genotypes_from_vcf_dataframe(t2t_mapped)\n",
    "\n",
    "VcfHandler.convert_vcf_dataframe_to_fasta_file(t2t_mapped, 'data/output_data/data_full_merged.T2T_mapped.fa', variant_sites_only=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
